### PR TITLE
[JBTM-3424] Make consistency in error messag printed in log and retur…

### PR DESCRIPTION
**Pull request by moulalis**
https://issues.redhat.com/browse/JBTM-3424

LRA
!MAIN !QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !AS_TESTS !TOMCAT !JACOCO

### CODE REVIEW

**Comment by moulalis** [comment]
TESTIT

**Comment by ochaloup** [comment]
TESTIT

**Comment by ochaloup** [comment]
Hello @moulalis.

Thank you for submitting you pull request and providing the fix. We appreciate it.
The fix starts with the good end but it does not cover the whole range of https://issues.redhat.com/browse/JBTM-3424. The snippet which is placed in the issue description is only an example how the code looks like. It does not mean that's the only place in our code where we consider the deficiency that the jira talks about. The point of the issue is to search the code (any java class in the Narayana lra submodule) and find the pattern and fix it.
On top of that, your fix provides the same (or similar) wording now. But what if the text of the error message changes? We would rather need to take the message and provide it directly to the `WebApplicationException`. Not to copying the text and hardcoding it to the class.
Something which is here: https://github.com/jbosstm/narayana/blob/master/rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java#L325
What I'm not sure right now is to verify how the `WebApplicationException` is logged at the coordinator if the message could not be logged twice now.

We are happy about you interest in Narayana LRA project. As my explanation could not contain all the details, please, join us at the our gitter channel here: https://gitter.im/jbosstm/community
We can discuss our perspectives.

Plus, excuse me but I'm going to close https://github.com/jbosstm/narayana/pull/1860 which points to 5.11 branch which is a maintenance branch where we do not plan to merge any LRA fixes.

**Review by xstefank — APPROVED** [review]

**Comment by xstefank** [comment]
Premature approve, dismiss it please. I didn't notice that the scope is bigger.

**Comment by ochaloup** [comment]
@xstefank thanks for the clarification. I think I do not have permission to dismiss the approval but I'm in contact with @moulalis and we discussed the scope of the issue. I think we are clear that there is needed some more changes to fix the issue (plus the current code change does not compile which needs to be fixed as well).

**Review by ochaloup — CHANGES_REQUESTED** [review]
Just explicitly marking the PR that it needs some adjustments (details in my comment above)

**Review by mmusgrov — APPROVED** [review]

**Comment by mmusgrov** [comment]
I also updated the JIRA indicating how to avoid this kind of discrepancy.

**Comment by mmusgrov** [comment]
I am copying my comment on the issue over here as well:

```
Can we also avoid duplicating the string for the error message, ie populate the exception message using the log message by returning a String from the message. An example is:

@Message(id = 25004, value = "LRA finished with an unexpected status code: %d, coordinator response '%s'")
String error_lraTerminationUnexpectedStatus(int status, String response);
`

**Comment by mmusgrov** [comment]
@moulalis what changes are you proposing in this PR. The only diff that I can see in the PR (including your recent force pushes) are the addition of a space character and the removal of a new line character. Each time an engineer pushes a new change to a PR it will trigger the consumption of CI resources in our Jenkins installation. Ondra has indicated what needs to be done and pushing commits that remove single white space characters is not a good use of engineering resource.

**Review by mmusgrov — CHANGES_REQUESTED** [review]
Please refer to [Ondra comment](https://github.com/jbosstm/narayana/pull/1859#issuecomment-865037902) to see what actions need to be done for this PR to qualify as a fix for JBTM-3424.

@moulalis I am adding the "On Hold" label while you decide whether or not you wish to pursue the fix. If you still want to fix it then I will remove the label when you propose a more comprehensive solution.

**Comment by moulalis** [comment]
> @moulalis what changes are you proposing in this PR. The only diff that I can see in the PR (including your recent force pushes) are the removal of a space character and the removal of a new line character. Each time an engineer pushes a new change to a PR it will trigger the consumption of CI resources in our Jenkins installation. Ondra has indicated what needs to be done and pushing commits that remove single white space characters is not a good use of engineering resources.

@mmusgrov It was my bad, sorry for it, I was checking why build is getting failed, '}' was missing but by any how I was suppose to push other PR and by mistake I did it here. It was all my bad.

**Comment by mmusgrov** [comment]
@moulalis Ah right okay, understood and fair enough, sorry if I was a bit terse.

**Comment by moulalis** [comment]
> Please refer to [Ondra comment](https://github.com/jbosstm/narayana/pull/1859#issuecomment-865037902) to see what actions need to be done for this PR to qualify as a fix for JBTM-3424.
> 
> @moulalis I am adding the "On Hold" label while you decide whether or not you wish to pursue the fix. If you still want to fix it then I will remove the label when you propose a more comprehensive solution.

I would be working on this and making the changes as Ondra suggested.

**Comment by mmusgrov** [comment]
> > Please refer to [Ondra comment](https://github.com/jbosstm/narayana/pull/1859#issuecomment-865037902) to see what actions need to be done for this PR to qualify as a fix for JBTM-3424.
> > @moulalis I am adding the "On Hold" label while you decide whether or not you wish to pursue the fix. If you still want to fix it then I will remove the label when you propose a more comprehensive solution.
> 
> I would be working on this and making the changes as Ondra suggested.

Okay, I will remove the hold label when you are ready with a more comprehensive fix, thanks.

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:501]: We don't normally make formatting changes (removing a line) amongst other changes.

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:633]: Insufficient parameters for format string.

File: rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java
Comment (mmusgrov) [inline rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java:137]: We don't allow formatting changes amongst functional changes (ie a blank line has been removed

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java
Comment (mmusgrov) [inline rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java:391]: You have missed all of the remaining data in the errMsg

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [reply rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:203]: The exception class, `e.getClass().getName()`, should already be reported in the logger call `error_invalidLraIdFormatToConvertToCoordinatorUrl`` .

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:330]: Insufficient parameters for format string. This mistake is also done elsewhere so I will not repeat this review comment but it needs fixing in all places.

**Review by mmusgrov — CHANGES_REQUESTED** [review]
@ochaloup please can you also take a look

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:348]: Incorrect usage of String.format. The syntax is defined in the javadoc for Formatter: https://docs.oracle.com/javase/8/docs/api/java/util/Formatter.html

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:354]: I would keep it simple and just use logMsg here.

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:437]: I would just use logMsg here.

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:508]: This statement serves no purpose

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:625]: This statement serves no purpose. I notice that you have done this in quite a few places in the PR so I will not mention further examples but it needs fixing.

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:760]: There is a mismatch here, in other places the exception message and the response to the client are the same text. For consistency I would keep them the same. Similarly in other places in this PR.

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java
Comment (mmusgrov) [inline rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java:391]: errMsg is not the same as the i18n log message (error_cannotGetStatusOfNestedLraURI) which it should be.

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java
Comment (mmusgrov) [inline rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java:695]: logMsg already contains the error message. Just use logMsg.

File: rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java
Comment (mmusgrov) [inline rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java:706]: logMsg already contains the error message. Just use logMsg.
You have done the same in other places so I will not repeat this comment but it needs changing throughout.

File: rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java
Comment (mmusgrov) [inline rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java:25]: The imports have been rearranged, we disallow formatting changes in PRs.

**Review by mmusgrov — CHANGES_REQUESTED** [review]
There are still a number of change requests needed on this PR.

**Comment by mmusgrov** [comment]
> Premature approve, dismiss it please. I didn't notice that the scope is bigger.

@xstefank you should have the requisite permissions to dismiss it - it would be setting a precedent if we do it for you.

**Comment by xstefank** [comment]
@mmusgrov I don't have access to the repo so I don't think I can dismiss reviews.

**Comment by mmusgrov** [comment]
@xstefank we don't have permission to dismiss your review either so let's just leave it as it is.

**Comment by moulalis** [comment]
@mmusgrov could you please review, I have moved the latest changes. 
Thanks

File: rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java
Comment (mmusgrov) [inline rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java:121]: You should avoid adding white space, I would keep the original spacing.

**Review by mmusgrov — CHANGES_REQUESTED** [review]
@moulalis Many of my review comments have either not been addressed or have not been responded to. If the author of the PR disagrees with a comment then there needs to be a follow up comment explaining why the change request has not been applied.
 
Also in my original review I pointed out some issues remarking that if a problem was repeated more than once I said I would not comment on all occurrences of the problem but that they all needed to be fixed. Here I am referring a number of examples of `String errorMsg = String.format("%s", logMsg);` which is a statement that serves no purpose.

And the following comment added by @ochaloup has not been responded to:

> ... it would be good to verify how the WebApplicationException is logged at the coordinator, could be the message is logged twice now (not sure)?

**Comment by moulalis** [comment]
@mmusgrov errorMsg and blank space removed. Thanks

**Comment by mmusgrov** [comment]
@moulalis I've put this PR on hold while you wait for input from @ochaloup and while you address the summary I provided in my last review.

File: rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java
Comment (ochaloup) [reply rts/lra/service-base/src/main/java/io/narayana/lra/logging/LraI18nLogger.java:53]: @mmusgrov I understood it in way that we can change the message but we can't change the `I18N` code. When we found the message is not precise we can change it to be comprehensible or better describing the intention. My understanding.

**Review by ochaloup — COMMENTED** [review]
To me it seems fine in general.

My point about `WebApplicationException` was that I was afraid if the exception is not logged at filter or at RestEasy level which would mean to log the same message twice or more times. Such behaviour would not seem to me as a nice.
I was thinking that would be good to verify - like via running some testcase what we have in TCK - how the message is logged when `WebApplicationException` is thrown.
@mmusgrov what do you think about having the exception being logged only by RestEasy (if it does so) and not printing errors in cases when the RestEasy does so?

**Comment by mmusgrov** [comment]
> To me it seems fine in general.
> 
> My point about `WebApplicationException` was that I was afraid if the exception is not logged at filter or at RestEasy level which would mean to log the same message twice or more times. Such behaviour would not seem to me as a nice.
> I was thinking that would be good to verify - like via running some testcase what we have in TCK - how the message is logged when `WebApplicationException` is thrown.
> @mmusgrov what do you think about having the exception being logged only by RestEasy (if it does so) and not printing errors in cases when the RestEasy does so?

So are you suggesting that @moulalis needs to run the TCK (or one of the TCKs tests that throw a WebApplicationException preceded by one of our log messages) and to then verify that there isn't more than one entry in the logs for the same problem.
I didn't quite get your point about RESTEasy logging since the developer would generally just enable LRA logging (and would not enable RESTEasy logging), is that what you meant?

> @mmusgrov I understood it in way that we can change the message but we can't change the I18N code. When we found the message is not precise we can change it to be comprehensible or better describing the intention. My understanding.

Okay, in that case we can accept this one particular change, but in general changing the log messages should be avoided since there is no place where a user may enter the i18n code to find out what it means (I do it by extracting the last 5 characters of the code and then I search the source base). I do have a further general comment about the error messages, if we are going to allow changes to the text then some of them really don't read very well and they could benefit from a rewording - perhaps we could do this before publishing the list of codes along with their meanings.

So @moulalis, I Ondra is saying that you need to run the TCK and verify whether or not any problems are logged twice, is that correct @ochaloup ?
You also need to address the comments that I left in [my last review of this PR](https://github.com/jbosstm/narayana/pull/1859#pullrequestreview-706643099) .

**Comment by ochaloup** [comment]
```
So are you suggesting that @moulalis needs to run the TCK (or one of the TCKs tests that throw a WebApplicationException preceded by one of our log messages) and to then verify that there isn't more than one entry in the logs for the same problem.
I didn't quite get your point about RESTEasy logging since the developer would generally just enable LRA logging (and would not enable RESTEasy logging), is that what you meant?

So @moulalis, Ondra is saying that you need to run the TCK and verify whether or not any problems are logged twice, is that correct @ochaloup ?
```

@mmusgrov yest, that' so. My perspective is that it would be good to not log the same problem twice. It would be good to verify it. The TCK or Narayana LRA integration test could help to verify the behaviour.
I mentioned RESTEAsy as it's the framework which manages the processing and the web exceptions are processed there and transformed to error codes. That's why I mentioned the RESTEasy and that it would be the place where the errors are printed (again).

**Comment by moulalis** [comment]
> So @moulalis, I Ondra is saying that you need to run the TCK and verify whether or not any problems are logged twice, is that correct @ochaloup ?
@ochaloup could you please run TCK for me as you suggested.

> You also need to address the comments that I left in [my last review of this PR](https://github.com/jbosstm/narayana/pull/1859#pullrequestreview-706643099) .
@mmusgrov I've already fixed and pushed my changes.

**Comment by ochaloup** [comment]
@moulalis 
What I meant was that I can help you to run the TCK on your machine. What you need is to first build Narayana codebase and then run the TCKs. Like

```
./build.sh install -Pcommunity -DskipTests
cd rts/lra/test/
mvn clean verify -Parq -pl tck
```

Or you can use nightly tests of the Narayana from our CI (Red Hat VPN access is needed though) here: http://narayanaci1.eng.hst.ams2.redhat.com/job/narayana/PROFILE=LRA,jdk=jdk8.latest,label=linux/

**Comment by moulalis** [comment]
> @moulalis
> What I meant was that I can help you to run the TCK on your machine. What you need is to first build Narayana codebase and then run the TCKs. Like
> 
> ```
> ./build.sh install -Pcommunity -DskipTests
> cd rts/lra/test/
> mvn clean verify -Parq -pl tck
> ```
> 
> Or you can use nightly tests of the Narayana from our CI (Red Hat VPN access is needed though) here: http://narayanaci1.eng.hst.ams2.redhat.com/job/narayana/PROFILE=LRA,jdk=jdk8.latest,label=linux/

Thanks @ochaloup I was trying say about CI from Narayana. I will try it in locally and check.

**Comment by ochaloup** [comment]
@moulalis @mmusgrov we discussed the behaviour with @moulalis and I agree that TCKs shows there should be no double logging of the event when `WebApplicationException` is thrown. I'm fine about my objection.
Thanks @moulalis for investigation.

**Review by ochaloup — APPROVED** [review]
The fix seems fine to me now.

**Comment by moulalis** [comment]
> @moulalis @mmusgrov we discussed the behaviour with @moulalis and I agree that TCKs shows there should be no double logging of the event when `WebApplicationException` is thrown. I'm fine about my objection.
> Thanks @moulalis for investigation.

Thanks @ochaloup attaching log.txt for 
[log.txt](https://github.com/jbosstm/narayana/files/6849624/log.txt)
reference.

**Comment by ochaloup** [comment]
@moulalis thanks for your adjustments.
Just wanted to clarify that I agree with request put by Mike. For merging this PR those have to be fixed first.

**Comment by moulalis** [comment]
> @moulalis thanks for your adjustments.
> Just wanted to clarify that I agree with request put by Mike. For merging this PR those have to be fixed first.

@ochaloup I've already fixed those. @mmusgrov I've double-checked might be I am not able to find which part is missing, could you please review once and let me know.
Thanks.

**Comment by mmusgrov** [comment]
https://github.com/jbosstm/narayana/pull/1859/files#r663920421
https://github.com/jbosstm/narayana/pull/1859/files#r663933880
https://github.com/jbosstm/narayana/pull/1859/files#r663945091

These three review comments either need to be acted upon or you need to justify why they are not applicable. It's fine to say that the reviewers comments are invalid but it's not okay to ignore them.

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [reply rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:752]: Note that ,although you appear to be adding extra information (uri.toASCIIString()) to the exception message, the uri is actually the same as the lraId which has already been added to logMsg, ie it adds no value. Therefore using the value in the variable logMsg is sufficient.

And I don't recall (without checking) what I was referring to when I said `Similarly in other places in this PR` so you should perform a quick sanity check to verify you haven't done something similar elsewhere.

**Comment by moulalis** [comment]
> https://github.com/jbosstm/narayana/pull/1859/files#r663920421
> https://github.com/jbosstm/narayana/pull/1859/files#r663933880
> https://github.com/jbosstm/narayana/pull/1859/files#r663945091
> 
> These three review comments either need to be acted upon or you need to justify why they are not applicable. It's fine to say that the reviewers comments are invalid but it's not okay to ignore them.

@mmusgrov Changes updated, for https://github.com/jbosstm/narayana/pull/1859/files#r663920421 I've asked if still required to use logMsg, didn't repeat the same to other two. Its seems like ignore all other message and make only logMsg consistent wherever logged. I made the changes as expected. 
Thanks

File: rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java
Comment (mmusgrov) [inline rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:749]: Part of what we discussed on this LRA is that the message in the exception response needs to correspond to the i18n log message. Here you are setting the exception message to be the same as the log message but you are then passing something different back to the caller, if you check the [javadoc for this exception](https://docs.oracle.com/javaee/7/api/javax/ws/rs/WebApplicationException.html#WebApplicationException-java.lang.String-javax.ws.rs.core.Response-) you should see what I mean.

I also said that the value in the uri variable is the same as the value in the lraId so the value in the response entity adds nothing, so this exception does not follow this advice. The intention is for the log message and the exception response to be consistent - you have followed this advice in other parts of this PR but we need consistency across the whole of the PR.

This looks like the only remaining issue on the PR so we are almost there!

**Comment by moulalis** [comment]
@mmusgrov Changes updated.
Thanks.

**Review by mmusgrov — APPROVED** [review]
The changes look good, I will merge them.

**Comment by mmusgrov** [comment]
I think the earlier approval from @ochaloup was contingent on resolving the other reviewer's comments which have now been fully addressed so I will merge this one.

**Pull request merged**
